### PR TITLE
[WIP] Port Catch2 v2.x to 3.x

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** The Catch2 v2.x testing framework is based on a header-only implementation. While flexible, this is known to increase compile times ([see here](https://github.com/catchorg/Catch2/blob/devel/docs/migrate-v2-to-v3.md#top)). Porting to the static library model of V3.x should help reduce compile times substantially for our testing suite.

**Description of the Change:** Migrate from Catch2 v2.x to v3.x

**Benefits:** Significantly reduced build times for our test suites.

**Possible Drawbacks:** More duplication and complexity for building and caching test-suite static library. 

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/src/tests/TestHelpers.hpp [hunk 1]
```diff
@@ -9,49 +9,78 @@
...
 #include <random>
 #include <string>
 #include <type_traits>
+#include <variant>
 #include <vector>
 
 namespace Pennylane {
 template <class T, class Alloc = std::allocator<T>> struct PLApprox {
-    const std::vector<T, Alloc> &comp_;
+    using param_t = std::variant<std::vector<T, Alloc>, T, std::monostate>;
```
Comment (chaeyeunpark) [inline pennylane_lightning/src/tests/TestHelpers.hpp:24]: Hi @mlxd, I indeed wanted to change the structure of this class using Catch2 v3's new matcher https://github.com/catchorg/Catch2/blob/devel/docs/matchers.md#writing-custom-matchers-new-style. In fact, I have implemented one for MrMustard in https://github.com/XanaduAI/MrMustard/blob/add_cpp_math/cpp/tests/test_helper.hpp#L35-L80. Can I jump in this PR and change this class? Or you can just remain the minimal working class here and I will make another PR.

File: pennylane_lightning/src/tests/TestHelpers.hpp [hunk 2]
```diff
@@ -9,143 +9,150 @@
...
 
-    explicit PLApprox(const std::vector<T, Alloc> &comp) : comp_{comp} {}
+template <class T> struct is_std_vector {
+    constexpr static bool value = false;
+};
+template <class T, class Alloc> struct is_std_vector<std::vector<T, Alloc>> {
+    constexpr static bool value = true;
+};
+template <class T>
+concept std_vector = is_std_vector<T>::value;
```
Comment (AmintorDusko) [inline pennylane_lightning/src/tests/TestHelpers.hpp:35]: Concepts are a very cool C++20 feature.

**Comment by mlxd** [comment]
All-in-all the current changes managed to shave about 1 minute off the Linux tests so far, but the wheels are still around 48 mins....
I think I will expand the scope of this Pr to improve build time and not just the test-suite upgrade (downgrade?)

**Comment by AmintorDusko** [comment]
> All-in-all the current changes managed to shave about 1 minute off the Linux tests so far, but the wheels are still around 48 mins.... I think I will expand the scope of this Pr to improve build time and not just the test-suite upgrade (downgrade?)

I think this change makes sense.

**Comment by mlxd** [comment]
Closing due to large divergences

**Pull request closed without merging**